File: cakk-domain/mysql/src/test/java/com/cakk/domain/facade/user/UserLikeFacadeTest.java
package com.cakk.domain.facade.user;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.IntStream;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;

import com.cakk.common.enums.ReturnCode;
import com.cakk.common.enums.Role;
import com.cakk.common.exception.CakkException;
import com.cakk.domain.base.FacadeTest;
import com.cakk.domain.mysql.entity.shop.CakeShop;
import com.cakk.domain.mysql.entity.user.User;
import com.cakk.domain.mysql.facade.user.UserLikeFacade;

class UserLikeFacadeTest extends FacadeTest {

	@InjectMocks
	private UserLikeFacade userLikeFacade;

	@Test
	@DisplayName("케이크 샵 기대돼요 동작을 성공한다")
	void likeCakeShop() {
		//given
		final User user = getUserFixture(Role.USER);
		final CakeShop cakeShop = getCakeShopFixture();

		//when
		userLikeFacade.likeCakeShop(user, cakeShop);

		//then
		assertThat(cakeShop.getLikeCount()).isEqualTo(1);
		final boolean liked = cakeShop.getShopLikes().stream().anyMatch(it -> it.getUser().equals(user));
		assertTrue(liked);
	}

	@Test
	@DisplayName("케이크 샵 기대돼요 동작이 50회 초과로 실패한다")
	void heartCakeShop2() {
		// given
		final User user = getUserFixture(Role.USER);
		final CakeShop cakeShop = getCakeShopFixture();

		IntStream.range(0, 50).forEach(i -> userLikeFacade.likeCakeShop(user, cakeShop));

		// when & then
		assertThrows(
			CakkException.class,
			() -> userLikeFacade.likeCakeShop(user, cakeShop),
			ReturnCode.MAX_CAKE_SHOP_LIKE.getMessage()
		);
	}
}

File: cakk-core/src/test/kotlin/com/cakk/core/facade/UserLikeFacadeTest.kt
package com.cakk.core.facade

import java.util.stream.IntStream

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe

import org.assertj.core.api.Assertions
import org.mockito.InjectMocks

import com.cakk.common.enums.ReturnCode
import com.cakk.common.enums.Role
import com.cakk.common.exception.CakkException
import com.cakk.core.common.annotation.TestWithDisplayName
import com.cakk.core.common.base.FacadeTest
import com.cakk.core.facade.user.UserLikeFacade
import com.cakk.domain.mysql.entity.shop.CakeShop
import com.cakk.domain.mysql.entity.user.User

internal class UserLikeFacadeTest : FacadeTest() {

	@InjectMocks
	private lateinit var userLikeFacade: UserLikeFacade

	@TestWithDisplayName("케이크 샵 기대돼요 동작을 성공한다")
	fun likeCakeShop() {
		//given
		val user: User = getUserFixture(Role.USER)
		val cakeShop: CakeShop = getCakeShopFixture()

		//when
		userLikeFacade.likeCakeShop(user, cakeShop)

		//then
		Assertions.assertThat(cakeShop.likeCount).isEqualTo(1)
		val liked: Boolean = cakeShop.shopLikes.stream().anyMatch { it.user == user }
		liked shouldBe true
	}

	@TestWithDisplayName("케이크 샵 기대돼요 동작이 50회 초과로 실패한다")
	fun heartCakeShop2() {
		// given
		val user: User = getUserFixture(Role.USER)
		val cakeShop: CakeShop = getCakeShopFixture()

		IntStream.range(0, 50).forEach { userLikeFacade.likeCakeShop(user, cakeShop) }

		// when
		val exception = shouldThrow<CakkException> {
			userLikeFacade.likeCakeShop(user, cakeShop)
		}

		// then
		exception.getReturnCode() shouldBe ReturnCode.MAX_CAKE_SHOP_LIKE
	}
}
